### PR TITLE
 Update: get reachable upload host using UploadReachability

### PR DESCRIPTION
**Pull request by wenboyu2**

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @wenboyu2 has signed the CLA. Thanks for the pull request!

File: src/api/PlainUpload.js [hunk 1]
```diff
@@ -21,15 +21,15 @@ class PlainUpload extends BaseUpload {
      * Handles a upload preflight success response
      *
      * @param {Object} data - Preflight success data
-     * @return {void}
+     * @return {Promise<>}
      */
-    uploadPreflightSuccessHandler = ({ upload_url }: { upload_url: string }) => {
+    uploadPreflightSuccessHandler = async ({ upload_url }: { upload_url: string }): Promise<> => {
```
Comment (priyajeet) [inline src/api/PlainUpload.js]: Sync to the latest, flow will complain about <> so will have to change it to something like Promise<any>

Comment (wenboyu2) [reply src/api/PlainUpload.js]: replace with Promise<*>

File: src/api/Base.js [hunk 2]
```diff
@@ -62,7 +65,7 @@ class Base {
      * @param {string} [options.uploadHost] - Upload host name
      * @return {Base} Base instance
      */
-    constructor(options: Options = {}) {
+    constructor(options: Options) {
```
Comment (wenboyu2) [inline src/api/Base.js]: removing `{}` as default, because flow complains about required attribute `token` missing

File: src/api/Base.js [hunk 3]
```diff
@@ -86,6 +90,17 @@ class Base {
...
     }
 
+    /**
+     * Update upload host with reachable url
+     *
+     * @return {Promise<*>}
+     */
+    async updateReachableUploadHost(): Promise<*> {
+        this.uploadHost = await this.uploadsReachability.getReachableUploadHost();
+        this.options.uploadHost = this.uploadHost;
```
Comment (tonyjin) [inline src/api/Base.js]: Any idea why we need to set `apiHost`, `cache` and `uploadHost` in `this.options` when we have all the data flattened in Base @priyajeet?

A little confusing to have to update it in two places.

File: src/api/PlainUpload.js [hunk 4]
```diff
@@ -21,15 +21,15 @@ class PlainUpload extends BaseUpload {
...
      */
-    uploadPreflightSuccessHandler = ({ upload_url }: { upload_url: string }) => {
+    uploadPreflightSuccessHandler = async ({ upload_url }: { upload_url: string }): Promise<*> => {
         if (this.isDestroyed()) {
             return;
         }
 
         // Make an actual POST request to the fast upload URL returned by pre-flight
-        this.makeRequest({
+        await this.makeRequest({
```
Comment (tonyjin) [inline src/api/PlainUpload.js]: TIL: https://eslint.org/docs/rules/no-return-await

File: src/api/uploads/UploadReachability.js [hunk 5]
```diff
@@ -4,31 +4,40 @@
...
+        this.xhr = new Xhr({
             // `id` is required for the preflight request, here it's set to `folder_0` because
             // the preflight request is for checking the availability of a host, not about a specific
             // upload attempt
             id: 'folder_0',
-            ...options
+            token
         });
+
+        const suffix: string = apiHost.endsWith('/') ? '2.0' : '/2.0';
```
Comment (tonyjin) [inline src/api/uploads/UploadReachability.js]: `endsWith` doesn't exist in IE11: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/endsWith

I'm not sure what our current policy of polyfills is in UI Elements @priyajeet - did we decide to recommend IE11 users include babel-polyfill?

File: src/api/uploads/UploadReachability.js [hunk 6]
```diff
@@ -116,12 +127,14 @@ class UploadsReachability extends Base {
...
-    handlePreflightResponse(response?: Object) {
+    handlePreflightResponse(response ? : Object) {
         if (!response) {
             return DEFAULT_HOSTNAME_UPLOAD;
         }
 
-        const { upload_url } = response;
+        const {
+            upload_url
+        } = response;
```
Comment (tonyjin) [inline src/api/uploads/UploadReachability.js]: Hmm?

File: src/api/uploads/UploadReachability.js [hunk 7]
```diff
@@ -4,31 +4,40 @@
...
+    DEFAULT_HOSTNAME_UPLOAD
+} from '../../constants';
+import Xhr from '../../util/Xhr';
+import type { Token } from '../../flowTypes';
 
 const CACHED_RESULTS_LOCAL_STORE_KEY = 'uploads-reachability-cached-results';
 
-class UploadsReachability extends Base {
+
+class UploadsReachability {
```
Comment (tonyjin) [inline src/api/uploads/UploadReachability.js]: Why'd you decide to no longer extend from `Base`?

Comment (wenboyu2) [reply src/api/uploads/UploadReachability.js]: - circular dependency. `UploadsReachability` is required in Base, for determining the upload host
- previously I extended from Base just for the `xhr` instance and `getBaseUrl`, both are trivial to be made available directly inside `UploadsReachability`

File: src/api/PlainUpload.js (hunk 4 above)
Comment (wenboyu2) [reply src/api/PlainUpload.js]: hmm maybe I should `return this.makeRequest()`

File: src/api/uploads/UploadReachability.js (hunk 5 above)
Comment (wenboyu2) [reply src/api/uploads/UploadReachability.js]: copied from Base

File: src/api/uploads/UploadReachability.js (hunk 6 above)
Comment (wenboyu2) [reply src/api/uploads/UploadReachability.js]: let me run through prettier again...

File: src/api/Base.js (hunk 3 above)
Comment (wenboyu2) [reply src/api/Base.js]: Actually, I think we don't need `this.options`. And `Xhr` doesn't accept apiHost, cache or uploadHost in the constructor `options` param. Fixing that

Comment (wenboyu2) [reply src/api/Base.js]: nvm, I'm not sure, keeping it as is for now and adding a comment

**Comment by wenboyu2** [comment]
@tonyjin Comment added. Please approve and merge if it looks fine

**Pull request merged**